pkg/stash/with_etcd_test.go

```diff
@@ -0,0 +1,81 @@
+package stash
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gomods/athens/pkg/storage"
+	"github.com/gomods/athens/pkg/storage/mem"
+	"golang.org/x/sync/errgroup"
+)
+
+// TestEtcdSingleFlight will ensure that 5 concurrent requests will all get the first request's
+// response. We can ensure that because only the first response does not return an error
+// and therefore all 5 responses should have no error.
+func TestEtcdSingleFlight(t *testing.T) {
+	endpointsStr := os.Getenv("ETCD_TEST_ENDPOINTS")
+	if len(endpointsStr) == 0 {
+		t.SkipNow()
+	}
+	strg, err := mem.NewStorage()
+	if err != nil {
+		t.Fatal(err)
+	}
+	ms := &mockEtcdStasher{strg: strg}
+	endpoints := strings.Split(endpointsStr, ",")
+	wrapper, err := WithEtcd(endpoints, strg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	s := wrapper(ms)
+
+	var eg errgroup.Group
+	for i := 0; i < 5; i++ {
+		eg.Go(func() error {
+			_, err := s.Stash(context.Background(), "mod", "ver")
+			return err
+		})
+	}
+
+	err = eg.Wait()
```

Comment:
is there a way to do this with a timeout? it'd be nice to avoid having integration tests hang forever if there's something wrong with an etcd server running in docker or something else. maybe this can be done by creating `eg` with a context that times out? I don't remember if that's how it works...

Reply:
@arschles nice suggestion, I'll add a timeout on the context of the Stash function, should work fine.